File: DataMaker.Sdk.Php/src/DmfReader.php
<?php

declare(strict_types=1);

namespace DataMaker\Sdk;

/**
 * Reader/verifier for the .dmf signed-bundle format
 * (DataMaker.Forms.Signing.DmfFormat). A .dmf is a ZIP of manifest.json (the
 * Ed25519-signed bytes), signature.bin, form.json, plus optional render assets
 * bound by SHA-256 in manifest.files[]. Verifies the manifest signature, the
 * form.json hash, and the FOBO attestation, then returns a sender-side
 * descriptor to submit against. Mirrors the Python/JS SDK readers.
 */
final class DmfReader
{
    private const FILE_MANIFEST  = 'manifest.json';
    private const FILE_SIGNATURE = 'signature.bin';
    private const FILE_FORM      = 'form.json';

    /** Hard cap on the inflated size of any single zip entry (zip-bomb guard). */
    private const MAX_ENTRY_BYTES = 16 * 1024 * 1024;

    /**
     * Parse + verify a .dmf. $verify gates the Ed25519 + form-hash checks
     * (disable only for trusted local fixtures). The FOBO attestation is
     * verified independently of $verify.
     */
    public static function read(string $dmfBytes, bool $verify = true): FormDescriptor
    {
        $tmp = tempnam(sys_get_temp_dir(), 'dmf');
        if ($tmp === false || file_put_contents($tmp, $dmfBytes) === false) {
            throw new DmfError('could not buffer the .dmf for reading');
        }

        $zip = new \ZipArchive();
        if ($zip->open($tmp) !== true) {
            @unlink($tmp);
            throw new DmfError('not a readable .dmf archive');
        }

        try {
            $manifestBytes = self::readEntry($zip, self::FILE_MANIFEST);
            $formBytes     = self::readEntry($zip, self::FILE_FORM);

            $manifest = json_decode($manifestBytes, true);
            if (!is_array($manifest)) {
                throw new DmfError('.dmf manifest.json is not valid JSON');
            }

            $signer = $manifest['signer'] ?? null;
            $signerPubB64 = is_array($signer) ? (string) ($signer['publicKey'] ?? '') : '';

            if ($verify) {
                if ($signerPubB64 === '') {
                    throw new DmfError('.dmf manifest.signer.publicKey is missing — cannot verify');
                }
                $signatureBytes = self::readEntry($zip, self::FILE_SIGNATURE);
                if (!Crypto::verifyEd25519($manifestBytes, $signatureBytes, $signerPubB64)) {
                    throw new DmfError('.dmf signature is invalid — bundle was tampered with or signer key is wrong');
                }
                self::verifyFormHash($manifest, $formBytes);
            }

            $form = json_decode($formBytes, true);
            if (!is_array($form)) {
                throw new DmfError('.dmf form.json is not valid JSON');
            }

            // FOBO attestation (optional) — its own chain, verified regardless of $verify.
            $fobo = null;
            if ($signerPubB64 !== '' && is_array($signer)) {
                $fobo = FoboTrustRoot::verify($signer['foboAttestation'] ?? null, $signerPubB64);
            }

            $recipient = $manifest['recipient'] ?? null;

            return new FormDescriptor([
                'formId'             => $form['id'] ?? '',
                'name'               => $form['name'] ?? '',
                'description'        => $form['description'] ?? null,
                'schemaVersion'      => $form['schemaVersion'] ?? 1,
                'submitPolicy'       => $form['submitPolicy'] ?? null,
                'recipientUserId'    => is_array($recipient) ? ($recipient['userId'] ?? null) : null,
                'recipientPublicKey' => is_array($recipient) ? ($recipient['publicKey'] ?? null) : null,
                'signer'             => is_array($signer) ? ['publicKey' => $signer['publicKey'] ?? null, 'identity' => $signer['identity'] ?? null] : null,
                'envelopeVersion'    => $manifest['envelopeVersion'] ?? 0,
                'fields'             => self::extractFields($form),
                'verified'           => $verify,
                'foboVerification'   => $fobo,
            ]);
        } finally {
            $zip->close();
            @unlink($tmp);
        }
    }

    /**
     * Flatten form.fields[] into submit-ready descriptors. `key` is the storage
     * name (what the receiver indexes on); `label` is the caption.
     *
     * @param array<string,mixed> $form
     * @return array<int,array<string,mixed>>
     */
    public static function extractFields(array $form): array
    {
        $out = [];
        foreach (($form['fields'] ?? []) as $f) {
            if (!is_array($f)) {
                continue;
            }
            $name = (string) ($f['name'] ?? '');
            $entry = [
                'id'       => $f['id'] ?? '',
                'key'      => $name,
                'name'     => $name,
                'label'    => $f['label'] ?? $name,
                'kind'     => $f['kind'] ?? '',
                'required' => (bool) ($f['required'] ?? false),
            ];
            if (!empty($f['description'])) {
                $entry['description'] = $f['description'];
            }
            if (!empty($f['placeholder'])) {
                $entry['placeholder'] = $f['placeholder'];
            }
            $choice = $f['choice'] ?? null;
            $choices = is_array($choice) ? ($choice['choices'] ?? null) : null;
            if (is_array($choices) && $choices !== []) {
                $entry['choices'] = array_map(static function ($c): array {
                    $v = $c['value'] ?? '';
                    return ['value' => $v, 'label' => $c['label'] ?? $v];
                }, $choices);
                $entry['allowCustom'] = (bool) ($choice['allowCustom'] ?? false);
            }
            $out[] = $entry;
        }
        return $out;
    }

    /** @param array<string,mixed> $manifest */
    private static function verifyFormHash(array $manifest, string $formBytes): void
    {
        $digest = Crypto::sha256Hex($formBytes);
        foreach (($manifest['files'] ?? []) as $f) {
            if (is_array($f) && ($f['path'] ?? null) === self::FILE_FORM) {
                if (strtolower((string) ($f['sha256'] ?? '')) !== $digest) {
                    throw new DmfError('.dmf form.json hash mismatch — bundle was tampered with');
                }
                return;
            }
        }
        throw new DmfError('.dmf manifest does not list form.json — bundle is malformed');
    }

    private static function readEntry(\ZipArchive $zip, string $path): string
    {
        $stat = $zip->statName($path);
        if ($stat === false) {
            throw new DmfError("bundle is missing required entry '{$path}'");
        }
        if (isset($stat['size']) && (int) $stat['size'] > self::MAX_ENTRY_BYTES) {
            throw new DmfError("entry '{$path}' exceeds the maximum allowed size");
        }
        $bytes = $zip->getFromName($path);
        if ($bytes === false) {
            throw new DmfError("bundle is missing required entry '{$path}'");
        }
        return $bytes;
    }
}

File: DataMaker.Sdk.Net/src/DataMaker.Sdk/DmfReader.cs
using System.IO.Compression;
using System.Text;
using System.Text.Json;

namespace DataMaker.Sdk;

/// <summary>
/// Reader/verifier for the .dmf signed-bundle format
/// (DataMaker.Forms.Signing.DmfFormat). A .dmf is a ZIP of manifest.json (the
/// Ed25519-signed bytes), signature.bin, form.json, plus optional render assets
/// bound by SHA-256 in manifest.files[]. Verifies the manifest signature and
/// that form.json matches its signed hash, then returns a descriptor to submit
/// against.
/// </summary>
public static class DmfReader
{
    private const string ManifestEntry = "manifest.json";
    private const string SignatureEntry = "signature.bin";
    private const string FormEntry = "form.json";
    private const string CompiledEntry = "compiled.json";
    private const string ElementCssEntry = "elementCss.json";
    private const string PaletteEntry = "palette.css";
    private const string FontsEntry = "fonts.css";

    /// <summary>
    /// Parse + verify a .dmf. <paramref name="verify"/> gates the Ed25519 +
    /// form-hash checks (disable only for trusted local fixtures).
    /// <paramref name="includeRenderBundle"/> also reads compiled/elementCss/
    /// palette/fonts entries for the ASP.NET renderer.
    /// </summary>
    public static FormDescriptor Read(byte[] dmfBytes, bool verify = true, bool includeRenderBundle = false)
    {
        ZipArchive zip;
        try
        {
            zip = new ZipArchive(new MemoryStream(dmfBytes), ZipArchiveMode.Read);
        }
        catch (Exception e)
        {
            throw new DmfException($"not a readable .dmf archive: {e.Message}");
        }

        using (zip)
        {
            var manifestBytes = ReadEntry(zip, ManifestEntry) ?? throw new DmfException(".dmf is missing manifest.json");
            var formBytes = ReadEntry(zip, FormEntry) ?? throw new DmfException(".dmf is missing form.json");

            using var manifestDoc = ParseJson(manifestBytes, "manifest.json");
            var manifest = manifestDoc.RootElement;

            if (verify)
            {
                var signatureBytes = ReadEntry(zip, SignatureEntry) ?? throw new DmfException(".dmf is missing signature.bin");
                VerifySignature(manifest, manifestBytes, signatureBytes);
                VerifyFormHash(manifest, formBytes);
            }

            using var formDoc = ParseJson(formBytes, "form.json");
            var form = formDoc.RootElement;

            var recipient = manifest.TryGetProperty("recipient", out var r) && r.ValueKind == JsonValueKind.Object ? r : (JsonElement?)null;
            var signer = manifest.TryGetProperty("signer", out var s) && s.ValueKind == JsonValueKind.Object ? s : (JsonElement?)null;

            // FOBO attestation (optional) — its own chain, verified against the
            // root pubkey independent of the manifest-signature `verify` flag.
            // Null when absent or it fails to verify.
            FoboAttestation? fobo = null;
            if (signer is { } sgFobo && GetString(sgFobo, "publicKey") is { } signerPub
                && sgFobo.TryGetProperty("foboAttestation", out var att) && att.ValueKind == JsonValueKind.Object)
            {
                fobo = FoboTrustRoot.Verify(att, signerPub);
            }

            RenderBundle? bundle = null;
            if (includeRenderBundle)
            {
                bundle = new RenderBundle
                {
                    FormJson = Encoding.UTF8.GetString(formBytes),
                    CompiledJson = ReadEntryText(zip, CompiledEntry),
                    ElementCssJson = ReadEntryText(zip, ElementCssEntry),
                    PaletteCss = ReadEntryText(zip, PaletteEntry),
                    FontsCss = ReadEntryText(zip, FontsEntry),
                };
            }

            return new FormDescriptor
            {
                FormId = GetString(form, "id") ?? "",
                Name = GetString(form, "name") ?? "",
                Description = GetString(form, "description"),
                SchemaVersion = GetInt(form, "schemaVersion") ?? 1,
                SubmitPolicy = GetString(form, "submitPolicy"),
                RecipientUserId = recipient is { } rec ? GetString(rec, "userId") : null,
                RecipientPublicKey = recipient is { } rec2 ? GetString(rec2, "publicKey") : null,
                Signer = signer is { } sg
                    ? new SignerInfo(
                        GetString(sg, "publicKey") ?? "",
                        sg.TryGetProperty("identity", out var id) && id.ValueKind == JsonValueKind.Object ? GetString(id, "email") : null,
                        sg.TryGetProperty("identity", out var id2) && id2.ValueKind == JsonValueKind.Object ? GetString(id2, "name") : null,
                        sg.TryGetProperty("identity", out var id3) && id3.ValueKind == JsonValueKind.Object ? GetString(id3, "company") : null)
                    : null,
                FoboVerification = fobo,
                EnvelopeVersion = GetInt(manifest, "envelopeVersion") ?? 0,
                Fields = ExtractFields(form),
                Verified = verify,
                RenderBundle = bundle,
            };
        }
    }

    private static void VerifySignature(JsonElement manifest, byte[] manifestBytes, byte[] signatureBytes)
    {
        if (!manifest.TryGetProperty("signer", out var signer) || GetString(signer, "publicKey") is not { } pub)
            throw new DmfException(".dmf manifest.signer.publicKey is missing — cannot verify");
        bool ok;
        try { ok = SealedBox.VerifyEd25519(manifestBytes, signatureBytes, pub); }
        catch (Exception e) { throw new DmfException(e.Message); }
        if (!ok)
            throw new DmfException(".dmf signature is invalid — bundle was tampered with or signer key is wrong");
    }

    private static void VerifyFormHash(JsonElement manifest, byte[] formBytes)
    {
        var digest = SealedBox.Sha256Hex(formBytes);
        if (!manifest.TryGetProperty("files", out var files) || files.ValueKind != JsonValueKind.Array)
            throw new DmfException(".dmf manifest does not list files — bundle is malformed");
        foreach (var f in files.EnumerateArray())
        {
            if (GetString(f, "path") == FormEntry)
            {
                if (GetString(f, "sha256") != digest)
                    throw new DmfException(".dmf form.json hash mismatch — bundle was tampered with");
                return;
            }
        }
        throw new DmfException(".dmf manifest does not list form.json — bundle is malformed");
    }

    private static IReadOnlyList<FieldDescriptor> ExtractFields(JsonElement form)
    {
        var list = new List<FieldDescriptor>();
        if (!form.TryGetProperty("fields", out var fields) || fields.ValueKind != JsonValueKind.Array)
            return list;

        foreach (var f in fields.EnumerateArray())
        {
            var name = GetString(f, "name") ?? "";
            List<ChoiceOption>? choices = null;
            var allowCustom = false;
            if (f.TryGetProperty("choice", out var choice) && choice.ValueKind == JsonValueKind.Object)
            {
                allowCustom = choice.TryGetProperty("allowCustom", out var ac) && ac.ValueKind == JsonValueKind.True;
                if (choice.TryGetProperty("choices", out var cs) && cs.ValueKind == JsonValueKind.Array)
                {
                    choices = new List<ChoiceOption>();
                    foreach (var c in cs.EnumerateArray())
                    {
                        var v = GetString(c, "value") ?? "";
                        choices.Add(new ChoiceOption(v, GetString(c, "label") ?? v));
                    }
                }
            }

            list.Add(new FieldDescriptor
            {
                Id = GetString(f, "id") ?? "",
                Key = name,
                Name = name,
                Label = GetString(f, "label") ?? name,
                Kind = GetString(f, "kind") ?? "",
                Required = f.TryGetProperty("required", out var req) && req.ValueKind == JsonValueKind.True,
                Description = GetString(f, "description"),
                Placeholder = GetString(f, "placeholder"),
                Choices = choices,
                AllowCustom = allowCustom,
            });
        }
        return list;
    }

    private static byte[]? ReadEntry(ZipArchive zip, string name)
    {
        var entry = zip.GetEntry(name);
        if (entry is null) return null;
        using var s = entry.Open();
        using var ms = new MemoryStream();
        s.CopyTo(ms);
        return ms.ToArray();
    }

    private static string? ReadEntryText(ZipArchive zip, string name)
    {
        var bytes = ReadEntry(zip, name);
        return bytes is null ? null : Encoding.UTF8.GetString(bytes);
    }

    private static JsonDocument ParseJson(byte[] bytes, string label)
    {
        try { return JsonDocument.Parse(bytes); }
        catch (Exception e) { throw new DmfException($".dmf {label} is not valid JSON: {e.Message}"); }
    }

    private static string? GetString(JsonElement el, string prop) =>
        el.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static int? GetInt(JsonElement el, string prop) =>
        el.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.Number && v.TryGetInt32(out var n) ? n : null;
}
